pkg/appmeshinject/inject.go

```diff
@@ -0,0 +1,216 @@
+		IgnoredIPs:         config.IgnoredIPs,
+		CpuRequests:        config.SidecarCpu,
+		MemoryRequests:     config.SidecarMemory,
+	}
+
+	// Proxyinit image or CNI
+	if isAppMeshCNIEnabled(pod) {
+		annotations := pod.GetAnnotations()
+		annotations[AppMeshEgressIgnoredIPsAnnotation] = config.IgnoredIPs
+		annotations[AppMeshEgressIgnoredPortsAnnotation] = initMeta.EgressIgnoredPorts
+		annotations[AppMeshPortsAnnotation] = initMeta.Ports
+		annotations[AppMeshSidecarInjectAnnotation] = "enabled"
+		annotations[AppMeshIgnoredUIDAnnotation] = AppMeshProxyUID
+		annotations[AppMeshProxyEgressPortAnnotation] = AppMeshProxyEgressPort
+		annotations[AppMeshProxyIngressPortAnnotation] = AppMeshProxyIngressPort
+		pod.SetAnnotations(annotations)
+	} else {
+		// inject proxyinit
+		proxyinit, err := renderInit(initMeta)
+		if err != nil {
+			return err
+		}
+		var initContainer corev1.Container
+		err = json.Unmarshal([]byte(proxyinit), &initContainer)
+		pod.Spec.InitContainers = append(pod.Spec.InitContainers, initContainer)
+	}
+
+	if config.EnableIAMForServiceAccounts && (pod.Spec.SecurityContext == nil || pod.Spec.SecurityContext.FSGroup == nil) {
+		dfsgroup := defaultFSGroup
+		if pod.Spec.SecurityContext == nil {
+			pod.Spec.SecurityContext = new(corev1.PodSecurityContext)
+		}
+		pod.Spec.SecurityContext.FSGroup = &dfsgroup
+	}
+
+	// Envoy and X-ray sidecars
+	sidecars, err := renderSidecars(sidecarMeta)
+
+	if err != nil {
+		return err
+	}
+	for _, v := range sidecars {
+		var container corev1.Container
+		err = json.Unmarshal([]byte(v), &container)
+		if err != nil {
+			return err
+		}
+		pod.Spec.Containers = append(pod.Spec.Containers, container)
+	}
+
+	// Has image pull secret
+	if config.EcrSecret {
+		ecrS := corev1.LocalObjectReference{Name: ecrSecret}
+		pod.Spec.ImagePullSecrets = append(pod.Spec.ImagePullSecrets, ecrS)
+	}
+
+	if config.EnableDatadogTracing {
+		var vol corev1.Volume
+		err = json.Unmarshal([]byte(renderDatadogConfigVolume()), &vol)
+		if err != nil {
+			return err
+		}
+		pod.Spec.Volumes = append(pod.Spec.Volumes, vol)
+		var container corev1.Container
+		datadogInit, err := renderDatadogInitContainer(config.DatadogAddress, config.DatadogPort)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal([]byte(datadogInit), &container)
+		if err != nil {
+			return err
+		}
+		pod.Spec.InitContainers = append(pod.Spec.InitContainers, container)
+	}
+
+	if config.EnableJaegerTracing {
+		var vol corev1.Volume
+		err = json.Unmarshal([]byte(renderJaegerConfigVolume()), &vol)
+		if err != nil {
+			return err
+		}
+		pod.Spec.Volumes = append(pod.Spec.Volumes, vol)
+		var container corev1.Container
+		jaegerInit, err := renderJaegerInitContainer(config.JaegerAddress, config.JaegerPort)
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal([]byte(jaegerInit), &container)
+		if err != nil {
+			return err
+		}
+		pod.Spec.InitContainers = append(pod.Spec.InitContainers, container)
+	}
```

Comment:
These if-statements look almost the same. Could they be extracted to an interface, then implemented in each sidecar file instead? Something like:
```go
type SidecarRenderer interface {
   renderInitContainer() (corev1.Container, error)
}

func (sc *SidecarConfig) renderInitContainer() corev1.Container, error {}
```
(Se comment below for more details)